src/widgets/panel/panel.js

```diff
@@ -0,0 +1,116 @@
+import { component } from '../../lib/suit';
+import Panel from '../../components/Panel/Panel';
+
+const suit = component('Panel');
+
+const renderer = ({ containerNode, cssClasses, templateProps }) => ({
+  renderingOptions,
+  hidden,
+}) => {
+  let bodyRef = null;
+
+  render(
+    <Panel
+      cssClasses={cssClasses}
+      hidden={hidden}
+      templateProps={templateProps}
+      data={renderingOptions}
+      onRef={ref => (bodyRef = ref)}
+    />,
+    containerNode
+  );
+
+  return { bodyRef };
+};
+
+const usage = `Usage:
+const widgetWithHeaderFooter = panel({
+  [ templates.{header, footer} ],
+  [ hidden ],
+  [ cssClasses.{root, noRefinementRoot, body, header, footer} ],
+})(widget);
+
+const myWidget = widgetWithHeaderFooter(widgetOptions)`;
+
+export default function panel({
+  templates = {},
+  hidden = () => false,
+  cssClasses: userCssClasses = {},
+} = {}) {
+  if (typeof hidden !== 'function') {
+    warn(
+      `The \`hidden\` option in the "panel" widget expects a function returning a boolean (received "${typeof hidden}" type).`
+    );
+  }
+
+  const cssClasses = {
+    root: cx(suit(), userCssClasses.root),
+    noRefinementRoot: cx(
+      suit({ modifierName: 'noRefinement' }),
+      userCssClasses.noRefinementRoot
+    ),
+    body: cx(suit({ descendantName: 'body' }), userCssClasses.body),
+    header: cx(suit({ descendantName: 'header' }), userCssClasses.header),
+    footer: cx(suit({ descendantName: 'footer' }), userCssClasses.footer),
+  };
+
+  return widgetFactory => (widgetOptions = {}) => {
+    const { container } = widgetOptions;
+
+    if (!container) {
+      throw new Error(
+        `[InstantSearch.js] The \`container\` option is required in the widget "${
+          widgetFactory.name
+        }".`
+      );
+    }
+
+    const defaultTemplates = { header: '', footer: '' };
+    const templateProps = prepareTemplateProps({ defaultTemplates, templates });
+
+    const renderPanel = renderer({
+      containerNode: getContainerNode(container),
+      cssClasses,
+      templateProps,
+    });
+
+    try {
+      const { bodyRef } = renderPanel({
+        renderingOptions: {
+          canRefine: false,
+        },
+        hidden: true,
+      });
+
+      const widget = widgetFactory({
+        ...widgetOptions,
+        container: getContainerNode(bodyRef),
+      });
+
+      return {
+        ...widget,
+        dispose() {
+          unmountComponentAtNode(getContainerNode(container));
+          widget.dispose();
+        },
+        render(options) {
+          const renderingOptions =
+            typeof widget.getRenderingOptions === 'function'
+              ? widget.getRenderingOptions(options)
+              : { canRefine: true };
+
+          renderPanel({
+            renderingOptions,
+            hidden: Boolean(hidden(renderingOptions)),
+          });
+          widget.render(options);
+        },
+      };
+    } catch (error) {
+      throw new Error(usage);
```

Comment:
This always reads as very abusive to me.
Correct me if I'm wrong, but the only errors the usage instructions are helpful for are type errors in props.
This means a lot of other errors simply swallowed. Doesn't this make it harder to debug?

Reply:
That's for sure, it's a pattern that exists in all widgets for now.

I'm working on a proposal for better error management in InstantSearch.js. This is outside of this PR's scope though.